syntax/grammar.mjs

```diff
@@ -0,0 +1,496 @@
+
+/* ----------------------------------------------- */
+/* Patterns consist of TextElements or Placeables. */
+let Pattern = defer(() =>
+    repeat1(
+        PatternElement)
+    // Flatten indented Placeables.
+    .map(flatten(1))
+    .map(into(FTL.Pattern)));
+
+let PatternElement = defer(() =>
+    either(
+        TextElement,
+        Placeable,
+        sequence(
+            break_indent.map(into(FTL.TextElement)),
+            Placeable)));
+
+let TextElement = defer(() =>
+    repeat1(
+        either(
+            text_char,
+            text_cont))
+    .map(to_string)
+    .map(into(FTL.TextElement)));
+
+let Placeable = defer(() =>
+    sequence(
+        char("{"),
+        between(
+            maybe(inline_space),
+            either(
+                // Order matters!
+                BlockExpression,
+                InlineExpression).as("expression")),
+        char("}"))
+    .map(to_object)
+    .map(into(FTL.Placeable)));
+
+let BlockExpression = defer(() =>
+    either(
+        SelectExpression,
+        VariantList));
+
+let InlineExpression = defer(() =>
+    either(
+        StringExpression,
+        NumberExpression,
+        CallExpression, // Must be before MessageReference
+        MessageAttributeExpression,
+        TermVariantExpression,
+        Identifier.map(into(FTL.MessageReference)),
+        TermIdentifier.map(into(FTL.MessageReference)),
+        ExternalArgument,
+        Placeable));
+
+/* ------------------ */
+/* Inline Expressions */
+let StringExpression = defer(() =>
+    quoted_text.map(into(FTL.StringExpression)));
+
+let NumberExpression = defer(() =>
+    number.map(into(FTL.NumberExpression)));
+
+let ExternalArgument = defer(() =>
+    ExternalIdentifier.map(into(FTL.ExternalArgument)));
+
+let CallExpression = defer(() =>
+    sequence(
+        Function.as("callee"),
+        char("("),
+        maybe(argument_list).as("args"),
+        maybe(inline_space),
+        char(")"))
+    .map(to_object)
+    .map(into(FTL.CallExpression)));
+
+let argument_list = defer(() =>
+    sequence(
+        Argument.as(),
+        repeat(
+            sequence(
+                char(","),
+                Argument.as())),
+        maybe(char(",")))
+    .map(flatten(2))
+    .map(to_array));
+
+let Argument = defer(() =>
+    between(
+        maybe(inline_space),
+        either(
+            NamedArgument,
+            InlineExpression)));
+
+let NamedArgument = defer(() =>
+    sequence(
+        Identifier.as("name"),
+        between(
+            maybe(inline_space),
+            char(":")),
+        either(
+            StringExpression,
+            NumberExpression).as("value"))
+    .map(to_object)
+    .map(into(FTL.NamedArgument)));
+
+let MessageAttributeExpression = defer(() =>
+    sequence(
+        Identifier.as("id"),
+        char("."),
+        Identifier.as("name"))
+    .map(to_object)
+    .map(into(FTL.AttributeExpression)));
+
+let TermAttributeExpression = defer(() =>
```

Comment:
Move this below `SelectorExpression`, and in particular out of inline expressions? As it's not one ;-)